README.md
The R package optimParallel 
===========================

The package provides a parallel versions of the L-BFGS-B optim method.
If the evaluation of the function fn takes more than 0.1 seconds,
optimParallel can significantly reduce the optimization time. For a p-parameter optimization,
the speed increase is about factor 1+2p when no analytic gradient is specified and
1+2p processor cores are available.

See the R Journal article https://doi.org/10.32614/RJ-2019-030 for more information.
It is also available as vignette.


R> vignette("optimParallel")





optimParallel/NEWS
- version 1.0-2:

- version 1.0-1: 
commit 6f389aa916dc8cbc2398b2ed3635713a4ecf1778
Author: Florian Gerber <florian.gerber@math.uzh.ch>
Date:   Tue Apr 21 15:26:15 2020 -0600
(1) pkg 'lbfgsb3' removed from suggest because the package has been archived.
(2) 'fn' and 'gr' can now have multiple arguments mixed with '...'.

- version 1.0: 
commit 373ce0334eb05a7c382bd95d086b83966ceb64e1
Author: Florian Gerber <florian.gerber@math.uzh.ch>
Date:   Mon Feb 24 17:51:45 2020 -0700
(1) bug fix: hessian has now correct sign if 'fnscale=-1'.
(2) Vignette and citation info updated.

- version 0.8: 
commit f0a8ed6e658cb481d549fd908d5640bf03abf394
Author: Florian Gerber <florian.gerber@math.uzh.ch>
Date:   Mon Feb 25 18:38:17 2019 +0100
(1) the mechanism to call 'fn' and 'gr' in parallel
    was improved to avoid unecessary copies of objects
    between processes in the cluster.
(2) the 'BFGS' and 'GC' methods are nolonger supported. 

- version 0.7-4: bug fix
commit 813a06269b6e777157fbe529bc62e38fc4d59de4
Author: Florian Gerber <florian.gerber@math.uzh.ch>
Date:   Mon Oct 15 22:40:37 2018 -0600
(1) 'fn' and 'gr' can now be functions from other packages 
    calling compiled code, e.g., 
    optimParallel(par=1, fn=dnorm, mean=3)

- version 0.7-3: bug fix
commit 01ebb9016000a7b087c01b190db0b7ab9d4778d6
Author: Florian Gerber <florian.gerber@math.uzh.ch>
Date:   Fri Jul 27 22:40:37 2018 -0600
(1) indexing 'par' by names inside 'fn' works now. 
    e.g. optimParallel(par=c(a=1), fn <- function(x) {x["a"]^2})
(2) using functions and method from loaded packages works now.
    'fn' and 'gr' have now .GlobalEnv as parent environment.
(3) the default value for the argument 'method' is now "L-BFGS-B".
(4) the help page ?optimParallel contains a new section 'Notes' 
    summarizing considerations for the practical use. 

- version 0.7-2: CRAN release
commit f46556cee870a5c13145ddfdf5f2e3998ea0b317
Author: Florian Gerber <florian.gerber@math.uzh.ch>
Date:   Tue May 1 12:43:35 2018 +0200

optimParallel/tests/testthat/test-issues.R
## rm(list=ls())
## library("testthat")
## library("optimParallel", lib.loc = "../../../lib/")
source("testsetup.R")

context("test-issues")

control <- structure(list(maxit = 10,
                          factr = 2.22044604925031e-16),
                     .Names = c("maxit","factr"))

FN1 <- function(par, sleep){
    Sys.sleep(sleep)
    sum(par^2)
}

GR1 <- function(par, sleep){
    Sys.sleep(sleep)
    2*par
}

test_that("optimParallel",{
    compareOptim(list(par=c(1,2,3), fn=FN1, gr=GR1, sleep=0,
                      
                      control=control),
                 verbose=verbose)
})

FN2 <- function(par, sleep){
    Sys.sleep(sleep)
    par["a"]^2+par["b"]^2
}
GR2 <- function(par, sleep){
    Sys.sleep(sleep)
    2*c(par["a"],par["b"])
}

FN3 <- function(par, sleep){
    Sys.sleep(sleep)
    par["a"]^2
}
GR3 <- function(par, sleep){
    Sys.sleep(sleep)
    2*c(par["a"])
}

test_that("optimParallel - named arguments",{
    compareOptim(list(par=c(a=1,b=2), fn=FN2, sleep=0,
                      control=control),
                 verbose=verbose)
    compareOptim(list(par=c(a=1,b=2), fn=FN2, gr= GR2, sleep=0,
                      control=control),
                 verbose=verbose)
    compareOptim(list(par=c(a=1), fn=FN3, sleep=0,
                      control=control),
                 verbose=verbose)
    compareOptim(list(par=c(a=1), fn=FN3, gr= GR3, sleep=0,
                      control=control),
                 verbose=verbose)
})


test_that("optimParallel - use compiled code from other packages",{
    compareOptim(list(par=c(a=1), fn=dnorm,
                      control=control),
                 verbose=verbose)
    compareOptim(list(par=c(a=1), fn=dnorm, mean=1,
                      control=control),
                 verbose=verbose)
})

FN4 <- function(par, a, sleep=0){
    Sys.sleep(sleep)
    sum(10*par^2) + a
}

GR4 <- function(par, b, sleep=0){
    Sys.sleep(sleep)
    b*2*par
}


test_that("optimParallel - fn and gr can have different aguments",{
    expect_equal(optimParallel(par=1, fn=FN4, gr=GR4,
                               a=1, b=10, sleep=0,
                               control=control)$par, 0)
})


test_that("optimParallel return correct sign of hessian if 'fnscale=-1'", {
    set.seed(13)
    x <- rnorm(1000, 5, 2)
    negll <- function(par, x) { -sum(dnorm(x=x, mean=par[1], sd=par[2], log=TRUE)) }
    posll <- function(par, x) {  sum(dnorm(x=x, mean=par[1], sd=par[2], log=TRUE)) }

    compareOptim(list(par=c(1,1), fn=negll, x=x,
                      control=control, hessian=TRUE),
                 verbose=verbose)

    compareOptim(list(par=c(1,1), fn=posll, x=x,
                      control=c(control, fnscale=-1), hessian=TRUE),
                 verbose=verbose)
})

test_that("fn can have normal and ... aguments", {
    fn <- function(par, data, ...)  par^2
    compareOptim(list(par=1, fn=fn, data=1:10,
                      control=control),
                 verbose=verbose)
})





optimParallel/tests/testthat/test-optimParallel.R
## rm(list=ls())
## library("testthat")
## library("optimParallel", lib.loc = "../../../lib/")
context("test-optimParallel")
source("testsetup.R")

x <- rnorm(1000, 2, .4)
f1 <- function(par, x){
#    print(par)
    -sum(dnorm(x, par[1], par[2], log=TRUE))
}
f2 <- function(par, y){
#    print(y)
    -sum(dnorm(y, par[1], par[2], log=TRUE))
}
f3 <- function(par){
    -sum(par)
}
f4 <- function(par){
   # cat("f:", par, "\n")
#    Sys.sleep(1)
    sum(par^2)
}
g4 <- function(par){
# cat("g:", par,"\n")
#    Sys.sleep(1)
    2*par
}

f5 <- function(par, a){
   # cat("f:", par, "\n")
#    Sys.sleep(1)
    sum(par^2)+a
}
g5 <- function(par, a){
   # cat("g:", par,"\n")
#    print(a)
                                        #    Sys.sleep(1)
    2*par
}

f6 <- function(par){
    par[1]^2+(1-par[2])^2
}

f7 <- function(x, ...){
    dots <- list(...)
    if(!identical(dots, list()))
        return(sum((x-dots[[1]])^2))
    sum(x^2)
}

f8 <- function(zz, x){
#    print(par)
    -sum(dnorm(x, zz[1], zz[2], log=TRUE))
}

test_that("optimParallel",{
    compareOptim(list(par=c(2,1), fn=f1, x=x, lower=c(-Inf,0.001),
                      control=list(factr=factr)), verbose=verbose)
    compareOptim(list(par=c(2,1), fn=f1, x=x,
                      lower=c(-Inf,0.001),
                      control=list(factr=factr)),
                 parallel=list(forward=TRUE),
                 verbose=verbose, tolerance=1e-2)
    
    compareOptim(list(par=c(12,100), fn=f1, x=x,
                      lower=c(-Inf,0.001),
                      control=list(factr=factr)), verbose=verbose)
    compareOptim(list(par=c(12,100), fn=f1, x=x,
                      lower=c(-Inf,0.001),
                      control=list(factr=factr)),
               parallel=list(forward=TRUE), tolerance=1e-2,
               verbose=verbose)

    compareOptim(list(par=c(12,100), fn=f2, y=x,
                      lower=c(-Inf,0.001),
                      control=list(factr=factr)), verbose=verbose)

    expect_error(optimParallel(par=c(12,100), fn=f2,
                               method = "L-BFGS-B",
                               lower=c(-Inf,0.001),
                               control=list(factr=factr)),
                 "argument \"y\" is missing, with no default")    
})


test_that("bounds",{
    compareOptim(list(par=c(2), fn=f3, 
                      upper = c(10),
                      control=list(factr=factr)),
                 verbose=verbose)
    compareOptim(list(par=c(2), fn=f3, 
                      upper = c(10),
                      control=list(factr=factr)),
                 parallel=list(forward=TRUE),
                 verbose=verbose)

    compareOptim(list(par=c(2,1), fn=f3, 
                      upper = c(10,15),
                      control=list(factr=factr)),
                 verbose=verbose)
    compareOptim(list(par=c(2,1), fn=f3, 
                      upper = c(10,15),
                      control=list(factr=factr)),
                 parallel=list(forward=TRUE),
                 verbose=verbose)

    compareOptim(list(par=c(12,100), fn=function(x) sum(x),
                      lower=c(14,-21),
                      control=list(factr=factr)), verbose=verbose)
    compareOptim(list(par=c(12,100), fn=function(x) sum(x),
                      lower=c(14,-21),
                      control=list(factr=factr)),
                 parallel=list(forward=TRUE), verbose=verbose)
})


test_that("ndeps",{
    compareOptim(list(par=c(2,1), fn=f1, x=x,
                      lower=c(-Inf,0.001),
                      control=list(factr=factr, ndeps=c(.1,.1))),
                 verbose=verbose)

    ## compareOptim(list(par=c(2,1), fn=f1,
    ##                   lower=c(-Inf,0.001),
    ##                   control=list(factr=factr, ndeps=c(.1,.1))),
    ##              parallel=list(forward=TRUE), tolerance=1e-3,
    ##              verbose=verbose)

    compareOptim(list(par=c(12,100), fn=f1, x=x,
                      lower=c(-Inf,0.001),
                      control=list(factr=factr, ndeps=c(.1,.1))),
                 verbose=verbose)
    ## compareOptim(list(par=c(12,100), fn=f1,
    ##                   lower=c(-Inf,0.001),
    ##                   control=list(factr=factr, ndeps=c(.1,.1))),
    ##              parallel=list(forward=TRUE), tolerance=1e-3,
    ##              verbose=verbose)
})


test_that("fnscale",{
    compareOptim(list(par=c(2,1), fn=f1, x=x,
                      lower=c(-Inf,0.001),
                      control=list(factr=factr, fnscale=1000)), verbose=verbose)
    compareOptim(list(par=c(2,1), fn=f1, x=x,
                      lower=c(-Inf,0.001),
                      control=list(factr=factr, fnscale=1000)),
                 parallel=list(forward=TRUE), tolerance=1e-3,
                 verbose=verbose)
})


test_that("parscale",{
    compareOptim(list(par=c(2,1), fn=f1, x=x,
                      lower=c(-Inf,0.001),
                      control=list(factr=factr,
                                   parscale=c(2,4))),
                 verbose=verbose)
    compareOptim(list(par=c(2,1), fn=f1, x=x,
                      lower=c(-Inf,0.001),
                      control=list(parscale=c(2,4), maxit=1)),
                 parallel=list(forward=TRUE), tolerance=1e-2,
                 verbose=verbose)
    
    compareOptim(list(par=c(2,1), fn=f1, x=x, lower=c(-Inf,0.001),
                      control=list(factr=factr, parscale=c(.2,4))), verbose=verbose)
    ## compareOptim(list(par=c(2,1), fn=f1, lower=c(-Inf,0.001),
    ##                   control=list(factr=factr, parscale=c(.2,4))),
    ##              parallel=list(forward=TRUE), tolerance=1e-3, verbose=verbose)
})

test_that("gradient",{
    compareOptim(list(par=c(2), fn=f4, gr=g4,
                      control=list(factr=factr)),
                 verbose=verbose)
    compareOptim(list(par=c(2,1), fn=f4, gr=g4,
                      control=list(factr=factr)),
                 verbose=verbose)
    compareOptim(list(par=c(3,2,1), fn=f4, gr=g4,
                      control=list(factr=factr)),
                 verbose=verbose)
})


test_that("... args",{
    compareOptim(list(par=c(2), fn=f5, gr=g5,
                      control=list(factr=factr), a=1),
                 verbose=verbose)
    compareOptim(list(par=c(2), fn=f5,
                      control=list(factr=factr), a=1),
                 verbose=verbose)
})


scratch/blog.R
## create files for blog entry
## https://florarblog.wordpress.com/2018/08/07/optimparallel-speed-up-your-optim-calls/

rm(list=ls())
set.seed(11)
library(optimParallel)

x <- rnorm(n = 1000, mean = 5, sd = 2)
negll <- function(par, x) -sum(dnorm(x = x, mean = par[1], sd = par[2], log = TRUE))
o1 <- optim(par = c(1, 1), fn = negll, x = x, method = "L-BFGS-B",
            lower = c(-Inf, 0.0001))
o1$par

cl <- makeCluster(detectCores()); setDefaultCluster(cl = cl)
o2 <- optimParallel(par = c(1, 1), fn = negll, x = x, method = "L-BFGS-B",
                    lower = c(-Inf, 0.0001))
identical(o1, o2)

o3 <- optimParallel(par = c(1, 1), fn = negll, x = x, method = "L-BFGS-B",
                    lower = c(-Inf, 0.0001), parallel=list(loginfo = TRUE))
print(o3$loginfo[1:3, ], digits = 3)

## make Figure 1
library(ggplot2); theme_set(theme_bw())
library(fields)
col2 <- tim.colors(nrow(o3$log)-2)
col2 <- rep("darkred", 100)
for(j in 1:(nrow(o3$log)-2)){
        grid <- expand.grid(par1=seq(.7, 7, length.out=100),
                            par2=seq(.7, 4, length.out=100),
                            z=NA)
        for(i in 1:nrow(grid))
            grid$z[i] <- negll(c(grid$par1[i], grid$par2[i]), x=x)
        col <- alpha(colorRampPalette(c("gray"))(55), 1)
        stroke <- 1.1
        shape <- 1
        p <- ggplot(data=data.frame(o3$log[1:j,]), aes(x=par1, y=par2)) +
        geom_contour(mapping=aes(x=par1, y=par2, z=z, color=..level..), data=grid,
                     bins=55) +
        geom_vline(xintercept=data.frame(o3$log)[j,"par1"], color=col2[j]) +
        geom_hline(yintercept=data.frame(o3$log)[j,"par2"], color=col2[j]) +
        geom_point(data=head(data.frame(o3$log),j-1), size =3, shape=3, color="darkred", alpha=1, stroke=1)+
        scale_color_gradientn(colours=col)+
        scale_x_continuous(expand = c(0,0), limits=range(grid$par1)) +
        scale_y_continuous(expand = c(0,0), limits=range(grid$par2)) +
        xlab(expression(mu)) + ylab(expression(sigma))+
         annotate("text", x = 6.17, y = 3.8, label = paste0("iteration:  "), size=8)  +     annotate("text", x = 6.9, y = 3.8, label = paste(j-1), size=8, hjust=1)+
                     annotate("text", x = 1, y = .85, label = paste0("start"), size=8) +
        theme(legend.position="none", axis.text=element_text(size=15),
              axis.title=element_text(size=16),
              plot.title = element_text(hjust = 0.5, size=20, face="bold"),
              panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.border = element_rect(colour = "black", fill=NA, size=1)) +
            ggtitle("Optimization path")


        if(j>1)
        p <- p+geom_segment(aes(xend=c(tail(par1, n=-1), NA), yend=c(tail(par2, n=-1), NA)),
                     arrow=arrow(length=unit(0.5,"cm"), type = "open", angle=30), size=.7,
                     color=col2[j], data=data.frame(o3$log[(j-1):j,])) +
            geom_segment(aes(xend=c(tail(par1, n=-1), NA), yend=c(tail(par2, n=-1), NA)), size=.7,
                     color=col2[1:j])
   
        ggsave(paste0("path_",sprintf("%02d", j),".png"), p,
               width = 8*.9*.9*1.3, height = .8*8*.9, device="png", dpi=200)
}

        grid <- expand.grid(par1=seq(.7, 7, length.out=100),
                            par2=seq(.7, 4, length.out=100),
                            z=NA)
        for(i in 1:nrow(grid))
            grid$z[i] <- negll(c(grid$par1[i], grid$par2[i]), x=x)
        col <- alpha(colorRampPalette(c("gray"))(55), 1)
        stroke <- 1.1
        shape <- 1
        p <- ggplot(data=data.frame(o3$log[1:j,]), aes(x=par1, y=par2)) +
        geom_contour(mapping=aes(x=par1, y=par2, z=z, color=..level..), data=grid,
                     bins=55) +
        geom_vline(xintercept=data.frame(o3$log)[j,"par1"], color=col2[j]) +
        geom_hline(yintercept=data.frame(o3$log)[j,"par2"], color=col2[j]) +
        geom_point(data=head(data.frame(o3$log),j-1), size =3, shape=3, color="darkred", alpha=1, stroke=1)+
        scale_color_gradientn(colours=col)+
        scale_x_continuous(expand = c(0,0), limits=range(grid$par1)) +
        scale_y_continuous(expand = c(0,0), limits=range(grid$par2)) +
        xlab(expression(mu)) + ylab(expression(sigma))+
             annotate("text", x = 6.17, y = 3.8, label = paste0("iteration:  "), size=8)  +     annotate("text", x = 6.9, y = 3.8, label = paste(j-1), size=8, hjust=1)+
                     annotate("text", x = 1, y = .85, label = paste0("start"), size=8) +
  theme(legend.position="none", axis.text=element_text(size=15),
              axis.title=element_text(size=16),
              plot.title = element_text(hjust = 0.5, size=20, face="bold"),
              panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.border = element_rect(colour = "black", fill=NA, size=1)) +ggtitle("Optimization path")


        if(j>1)
        p <- p+
            geom_segment(aes(xend=c(tail(par1, n=-1), NA), yend=c(tail(par2, n=-1), NA)), size=.7,
                     color=col2[1:j])



for(k in (nrow(o3$log)):(nrow(o3$log)+2))
    ggsave(paste0("path_",sprintf("%02d", k),".png"), p,
           width = 8*.9*.9*1.3, height = .8*8*.9, device="png", dpi=200)

system("convert -delay 100 path_*.png test.gif")



scratch/plot.optimParallel.R
plot(1:100, col=colorRampPalette(c("#332288", "#88CCEE", "#44AA99",
                "#CC6677", "#AA4499", "#999933", "#117733"))(100),
     pch=19, cex=6)
plot(1:8, col=colorRampPalette(c("#332288", "#44AA99", "#88CCEE", 
                                 "#AA4499", "#882255", "#999933", "#117733"))(8),
     pch=19, cex=6)


plot.optimParallel <- function(oo, what=c("par", "gradient"), legendPosition="topleft", legendText=NULL, lwd=1.5, main="", ...){
    if(is.null(oo$loginfo))
        stop("Log information is missing. Call 'optimParallel()' with argument 'parallel=list(loginfo=TRUE)' to generate the necessary log information.")
        
    opar <- par("xpd", "mai")
    on.exit(par(opar))
    par(mai=c(.8,1,.8,1))
    what <- match.arg(what)
    if(what=="par"){
        indexPar <- 1:(ncol(oo$loginfo)/2-1)+1
        ylab <- "pars"
    } else {
        indexPar <- 1:(ncol(oo$loginfo)/2-1)+1+ncol(oo$loginfo)/2
        ylab <- "gradients"
    }
    nPar <- length(indexPar)
    indexFn <- (ncol(oo$loginfo)/2)+1
    colPar <- colorRampPalette(c("#332288", "#44AA99", "#88CCEE", 
                                 "#AA4499", "#882255", "#999933", "#117733"))(length(indexPar))
    colFn <- "black"

    if(is.null(legendText))
        legendText <- colnames(oo$loginfo)[c(indexPar, indexFn)]
    else if(length(legendText) != nPar+1)
        stop("argument 'legendText' has to be of length #parameters + 1.")
    matplot(oo$loginfo[,indexPar], type="l", lty=1, yaxt="n", col=colPar,
            xlab="", ylab=ylab, panel.first={grid(col="#8080cc", lty=1, lwd=.3)}, lwd=lwd, main=main, ...)
    mtext("step", side=1, line=2)
    axis(2, las=2)
    par(new=TRUE)
    plot(oo$loginfo[,1], oo$loginfo[,indexFn], xlab="", xaxt="n", yaxt="n", ylab="", type="l", col=colFn, lty=2, lwd=lwd, ...)
    axis(4, las=2, col=colFn)
    corners  <-  par("usr") # Gets the four corners of plot area (x1, x2, y1, y2)
    par(xpd = TRUE) # Draw outside plot area
    text(x = corners[2]*1.15, y = mean(corners[3:4]), "fn()", srt = 270)
    par(xpd = FALSE)
    legend(legendPosition, legend=legendText, col=c(colPar, colFn), lty=c(rep(1, nPar), 2) , bg="white", lwd=lwd, ...)
}
## check for global variables
codetools::findGlobals(plot.optimParallel, merge=FALSE)$variables


scratch/scratch.R
rm(list = ls())
detach("package:optimParallel", unload=TRUE)
system("make lib")
library("optimParallel", lib.loc = "lib/")
## todo:
## - other optim methods
## - cluster make and stop automatically on.exit()
## - set Cores, 

## for Linux like platforms:
## -------------------------
cl <- makeCluster(3)
setDefaultCluster(cl)


## passing objects to fn/gr (see section 'Notes')
## ----------------------------------------------
a <- 10
fn <- function(par, b) sum((par-a-b)^2) 

## approach 1:
clusterExport(cl, "a")
optimParallel(par=1, fn=fn, b=1)

## approach 2 (recommended):
## rewrite 'fn' such that all necessary objects
## are passed as arguments
fn <- function(par, a, b) sum((par-a-b)^2) 
optimParallel(par=1, fn=fn, a=20, b=1)


negll <- function(par=0, sleep, verbose=FALSE){
    if(verbose)
        cat(par, "\n")
    Sys.sleep(sleep)
    -sum(dnorm(x=a, mean=par["mean"], sd=par["sd"], log=TRUE))
}
set.seed(13); a <- rnorm(1e3, 5, 2)


negll(1, sleep=0)

cl <- makeCluster(3)
setDefaultCluster(cl)
o1 <- optimParallel(par=c(mean=1, sd=1), fn=negll, 
                    verbose=TRUE,
                    sleep=0,
                    lower=c(-Inf), method="L-BFGS-B")

clusterExport(cl=cl,  )

system.time(o2 <- optim(par=c(mean=1,sd=1), fn=negll, verbose=TRUE, sleep=0,
                        lower=c(-Inf, .0001), method="L-BFGS-B"))



     optimParallel(par=c(mean=1,sd=1), fn=dnorm, mean=1, sd=1,
                   method = "L-BFGS-B", lower=c(-Inf, .0001))

     ## it can be faster to reuse an existing cluster multiple times.
     cl <- makeCluster(detectCores())
     optimParallel(par=c(1,1), fn=negll, x=x,
                   method = "L-BFGS-B", lower=c(-Inf, .0001),
                   parallel_args = list(cl=cl))
     
     ## the cluster can also be set in options().
     options(optimParallel.cl=cl)
     optimParallel(par=c(1,1), fn=negll,
                   method = "L-BFGS-B", lower=c(-Inf, .0001))
     
     ## when no longer used, the cluster can be stopped.
     stopCluster(cl); options(optimParallel.cl=NULL)

cl <-      makeForkCluster(nnodes = getOption("mc.cores", 2L))

a <- parLapply(cl=cl, 1:4, cat)

detectCores()                             # available cores
options(mc.cores=detectCores())           # set number of cores
options(optimParallel.lapply="mclapply")  # used by default

optimParallel(par=c(1,1), fn=neg2ll,
              method = "L-BFGS-B", lower=c(-Inf, .0001))

optimParallel(par=c(1,1), fn=neg2ll,
              verbose=TRUE, sleep=.5, # args to neg2ll()
              method = "L-BFGS-B", lower=c(-Inf, .0001),
              control= list(factr = .01/.Machine$double.eps),
              parallel_args=list(log=TRUE))
## each step invokes 5 parallel calls to neg2ll()


tt <- optimParallel(par=c(1,1), fn=neg2ll,
              verbose=TRUE, sleep=.5, 
              method = "L-BFGS-B", lower=c(-Inf, .0001),
              parallel_args=list(log=TRUE, one_sided=TRUE))
## each step invokes 3 parallel calls to neg2ll()

tt <- optimParallel(par=c(1,1), fn=neg2ll,
                    verbose=TRUE, sleep=.5, 
                    method = "L-BFGS-B", 
                    control= list(factr = .01/.Machine$double.eps),
                    parallel_args=list(log=TRUE, one_sided=TRUE))

## each step invokes 3 parallel calls to neg2ll()

optim(par=c(1,1), fn=neg2ll,
      verbose=TRUE, sleep=.5, 
      method = "BFGS", 
      control= list(factr = .01/.Machine$double.eps))






## Use option cl.cores to choose an appropriate cluster size.
cl <- makeCluster(2, outfile="")
verbose=TRUE
optimParallel(par=c(1,1), fn=neg2ll, x=x,
              verbose=TRUE, sleep=0, 
              method = "L-BFGS-B", lower=c(-Inf, .0001),
              parallel_args=list(lapply="parLapply", cl=cl))


system.time(optimParallel(par=c(1,1), fn=neg2ll, x=x,
              verbose=FALSE, sleep=.3, 
              method = "L-BFGS-B", lower=c(-Inf, .0001),
              parallel_args=list(lapply="lapply")))


neg2ll <- function(par, sleep=0, verbose=FALSE){
    if(verbose)
        cat(par, "\n")
    Sys.sleep(sleep)
    -sum(dnorm(x=x, mean=par[1], sd=par[2], log=TRUE))
}
x <- rnorm(1000, 5, 2)


options(optimParallel.lapply="parLapply")
cl <- makeCluster(detectCores())

optimParallel(par=c(1,1), fn=neg2ll,
              x=x, # it is required to pass all data used by fn 
              method = "L-BFGS-B", lower=c(-Inf, .0001),
              parallel_args=list(cl=cl))

optimParallel(par=c(1,1), fn=neg2ll,
              x=x, verbose=TRUE, sleep=.5,
              method="L-BFGS-B", lower=c(-Inf, .0001),
              control=list(factr=.01/.Machine$double.eps),
              parallel_args=list(cl=cl))
## note: printing to screen does not work with clusters

optimParallel(par=c(1,1), fn=neg2ll,
              x=x, verbose=TRUE, sleep=.5,
              method ="L-BFGS-B", lower=c(-Inf, .0001),
              control=list(factr=.01/.Machine$double.eps),
              parallel_args=list(cl=cl, one_sided=TRUE))

stopCluster(cl)

## optim and spatial prediction on a transect
n <- 100
loc <- runif(n)
distmat <- as.matrix(dist(loc))
expcov <- function(h, par){
    par[2] * exp(-h/par[1])
}
Sigma <- expcov(distmat, c(2,4))
data <- c(rmvnorm.spam(1, Sigma=as.spam(Sigma)))
plot(loc, data)









test <- function(...){
    args <- list(...)
    if(any(names(args) %in% c("cl", "X", "x", "fun", "mc.preschedule", "mc.set.seed", "mc.silent", 
                              "mc.cores", "mc.cleanup", "mc.allow.recursive"))
       stop("arguments passer though '...' cannot be named: \n
'cl', 'X', 'x', 'fun', 'mc.preschedule', 'mc.set.seed', 'mc.silent',\n
'mc.cores', 'mc.cleanup', 'mc.allow.recursive'. ")
}








negll <- function(par, data, sleep=0, cl, verbose=FALSE){
        if(verbose)
           cat(par, "\n")
        Sys.sleep(sleep)
        -sum(dnorm(x=data, mean=par[1], sd=par[2], log=TRUE))
     }
     set.seed(13); data <- rnorm(1000, 5, 2)
     
     optimParallel(par=c(1,1), fn=negll, data=data, cl=1,
                   method = "L-BFGS-B", lower=c(-Inf, .0001))
     
     optimParallel(par=c(1,1), fn=negll, x=x,
                   verbose=TRUE, sleep=.5, # args to negll()
                   method="L-BFGS-B", lower=c(-Inf, .0001),
                   control=list(factr=.01/.Machine$double.eps))
     ## each step invokes 5 parallel calls to negll()
     
     optimParallel(par=c(1,1), fn=negll, x=x,
                   verbose=TRUE, sleep=.5,
                   method ="L-BFGS-B", lower=c(-Inf, .0001),
                   control=list(factr=.01/.Machine$double.eps),
                   parallel_args=list(one_sided=TRUE))
     ## each step invokes 3 parallel calls to negll()
     
     ## store and access the log file:
     o <- optimParallel(par=c(1,1), fn=negll, x=x, method = "L-BFGS-B",
                        lower=c(-Inf, .0001), parallel_args=list(log=TRUE))
     o$log
     
     
     ## default values of the argument 'paralle_args'
     ## ---------------------------------------------
     options("optimParallel.one_sided", "optimParallel.log",
             "optimParallel.lapply", "optimParallel.cl")
     ## they can be changed via, e.g.,
     # options(optimParallel.log=TRUE)
     
     
     ## usage of parLapply (default on Windows)
     ------------------------------------------
     options(optimParallel.lapply="parLapply")
     ## optimParallel() makes a cluster and stops it on exit.
     optimParallel(par=c(1,1), fn=negll, x=x,
                   method = "L-BFGS-B", lower=c(-Inf, .0001))
     
     ## it can be faster to reuse an existing cluster multiple times.
     cl <- makeCluster(detectCores())
     optimParallel(par=c(1,1), fn=negll, x=x, 
                   method = "L-BFGS-B", lower=c(-Inf, .0001),
                   parallel_args = list(cl=cl))
     
     ## the cluster can also be set in options().
     options(optimParallel.cl=cl)
     optimParallel(par=c(1,1), fn=negll, x=x,
                   method = "L-BFGS-B", lower=c(-Inf, .0001))
     
     ## when no longer used, the cluster can be stopped.
     stopCluster(cl); options(optimParallel.cl=NULL)
     ## End(Not run)




test  <- function(a, b){
    list(...)
    
}

where <- function(name, env = parent.frame()) {
  if (identical(env, emptyenv())) {
    # Base case
    stop("Can't find ", name, call. = FALSE)
    
  } else if (exists(name, envir = env, inherits = FALSE)) {
    # Success case
    env
    
  } else {
    # Recursive case
    where(name, parent.env(env))
    
  }
}

    
renameargs <- function(fun, forbidden="f1"){
    names_in <- names(formals(fun))
    rena <- names_in %in% forbidden
    if(!(any((rena))))
        return(fun)
    names_new <- ifelse(rena, paste0(names_in,"_AVeryUnliklyName"), names_in)
    
    args <- lapply(seq_along(names_in),
                   function(x){ eval(parse(text=paste0("quote(",names_new[x],")"))) })
    names(args) <- names_in

    env_A <- new.env(parent=parent.frame(1))
    assign("fun", fun, envir=env_A)
    fout <- function(a){
        names_in_A <- names_in
        names_new_A <- names_new
        rm(names_in, names_new)
        print(names_in_A)
        for(i in seq_along(names_in))
            assign(names_in[i], eval(parse(text=names_new[i])), envir=env_A)
        ls( envir=env_A)
        rm(list = c("what", "args", "quote", "envir"))
        print(get("args", envir=env_A))
        do.call(what="fun", args=args, quote=FALSE, envir=env_A)

    }
    args_fout <- vector("pairlist",length(args))
    names(args_fout) <- names_new
    formals(fout) <- args_fout
    fout
}


test <- function(names_in, args){
    names_in+args
}

t1 <- renameargs(test, "names_in")
t1(a=2, args=3)
t1(a=2, args=3)
t1(names_in_AVeryUnliklyName=2, args=3)

f <- function(x){
    y <- x+1
    y
}

f <- function(x) x


tt <- gsub("x", "y", as.character(body(f)))
body(f) <- quote(tt)

body(f) <- parse(text=gsub("x", "y", as.character(body(f))))
    
renameArguments <- function(f, argnames="x", suffix="_AVeryUnliklyName"){
    stopifnot(is.function(f),
              is.character(argnames),
              is.character(suffix), length(suffix)==1)
    argnames <- argnames[argnames%in%names(formals(f))]
    bb <- as.character(body(f))
    if(bb[1]=="{")
        bb <- bb[2:length(bb)]
    bb <- paste0(bb, collapse="; ")
    bb <- paste0(c("{", bb, "}"), collapse="")

    for(i in seq_along(argnames))
        bb <- gsub(argnames[i], paste0(argnames[i], suffix, collapse=""), bb)
    body(f) <- parse(text=bb)

    for(i in seq_along(names(formals(f))))
        names(formals(f))[i][names(formals(f))[i]%in%argnames[i]] <- paste0(argnames[i], suffix, collapse="")
    f
}

    change(f, "x")


    tf <- function(a=1, b, c, d){
        x <- a+b
        y <- c+d
        x+y
    }
    tfo <- change(tf, c("a", "x"))


require(parallel)
cl <- makeCluster(2)
y <- 1
f <- function(x, y){
    x+y
}
args <- list(x=1)

gen <- function(f, args){
    for(i in seq_along(args)){
        assign(names(args)[i], args[[i]])
    }
    ff <- function(){}
    body(ff) <- body(f)
    update_fistargument <- function(value){
        assign(names(args)[1], value, inherits=TRUE)
    }
    list(f=ff, update_fistargument=update_fistargument)
}

integrates_args <- function(f, args){
    for(i in seq_along(args)){
        assign(names(args)[i], args[[i]])
    }
    ff <- function(){}
    body(ff) <- body(f)
    ff
}



par <- matrix(1:3)
getList <- function(par, args, f){
    lapply(1:nrow(par), function(x){
        args[[1]] <- par[x]
        gen(f=f, args)
    })
}

getList(matrix(1:3), list(x=1,y=2), f)

    
parLapply()
mclapply(1:3, FUN=function(x, FUN) {sum(x)}, FUN=1 )


    
